### DISCUSSION TITLE
Selection of CNNs for the rover model

### DISCUSSION THREAD

**Discussion by dietmarwo** [esa/SpOC2#22] [Morphing Rovers Q&A] [answered]
The Controller model essentialy transforms a rover view / heightmap into an action the rover performs. This task is quite similar to image recognition, so the application of CNNs is straightforward. Recently an alternative to CNNs raised in popularity: The Vision Transformer based models. 

In https://lightning.ai/docs/pytorch/stable/notebooks/course_UvA-DL/11-vision-transformer.html Transformers are compared to CNNs analyzing their strengths/disadvantages. From this comparison it is concluded:

- a CNN architecture integrates the concept of distance in an image: two pixels that are close to each other are more related than two distant pixels. Local patterns are combined into larger patterns, until we perform our classification prediction. All those aspects are inductive biases of a CNN. 

- a Transformer does not know which two pixels are close to each other, and which are far apart. It has to learn this information solely from the sparse learning signal of the classification task. 

- long-distance relations between local patterns can be difficult to process in CNNs, while in Transformers, all patches have the distance of one.

Are you sure about your preference of CNNs over Transformers for the rover model? What were the exact arguments motivating your decision? Is it because long-distance relations  between local patterns are less important than the fact that closer pixels have a stronger relation between each other?

Another model related issue:

        rb += [1,1,6,6,6,6,4]

forbids relu for the last activation, which probably is a bad idea. Is there any motivation for this decision?

**Comment by dodo47** [comment, marked as answer]
Hi dietmarwo,

that's a brilliant question :) There are actually several valid options for the controller that could be explored, so we are happy to discuss this in more detail after the competition ended!

In general, the terrain images are rather small, and what's important for the rover are features of local terrain patches. So the way we phrased it, the rover input is similar to an image regression problem, for which CNN is a good base model. Transformers definitely have the benefit of being able to model long-distance relations, but they also require more data to be trained competitively. So in this case, we decided to go with the standard approach + a small network to solve this task.

We are definitely interested to hear about the experience participants had in this regard though :)

For the last question: the last layer is recurrent, so to avoid exploding solutions that lead to NaN as a score, we only allow saturating activation functions in this layer. This way, every valid chromosome returns a valid (i.e., not NaN) score.

Cheers,
Dominik

**Reply by dietmarwo** [reply]
>This way, every valid chromosome returns a valid (i.e., not NaN) score.

Seems a valid motivation. relu works suprizingly well here nevertheless - beside the problem you mentioned. 

Regarding the Transformers: I didn't had time yet for experiments, although they may be interesting to perform. 
I am more focussed on getting good scores in all three tasks for now - and on GTOC12 starting next monday.
